### PR TITLE
[feature] #3641: Human-readable permission token payload

### LINKED ISSUE #3641: Make permission token payload human readable in `genesis.json`
So the bad part is that we loose readability in genesis. Just mentioning

_Originally posted by @Arjentix in https://github.com/hyperledger/iroha/pull/3625#discussion_r1239551590_

### PR DESCRIPTION
**Pull request by Arjentix**
## Description

**Scale** codec for Permission Token Payloads was replaced with **JSON**.

That's the most congruent way to fix the problem of readability. Permission tokens aren't used frequently and their main purpose is to be transparent for users. So their payload should be always readable.

Also did refactor a bit.

### Linked issue

- Closes #3641  

### Checklist

- [x] I've read `CONTRIBUTING.md`
- [x] I've used the standard signed-off commit format (or will squash just before merging)
- [x] All applicable CI checks pass (or I promised to make them pass later)
- [ ] (optional) I've written unit tests for the code changes
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: data_model/src/permission.rs
Comment (Erigara) [inline data_model/src/permission.rs:103]: Why not to convert`Value` to the `String` inside constructor? Because common pattern for this struct is `.to_string()` before actual use which might be costly.

**Review by Erigara — DISMISSED** [review]

File: wasm/validator/src/permission.rs
Comment (mversic) [inline wasm/validator/src/permission.rs:53]: Do we need `Value` and `Param` variants anymore?

File: wasm/validator/Cargo.toml
Comment (mversic) [inline wasm/validator/Cargo.toml:22]: do we still need SCALE?

**Review by mversic — CHANGES_REQUESTED** [review]

File: client/tests/integration/smartcontracts/validator_with_custom_token/src/lib.rs
Comment (mversic) [reply client/tests/integration/smartcontracts/validator_with_custom_token/src/lib.rs:121]: no big deal. Especially if you apply [this](https://github.com/hyperledger/iroha/pull/3775#discussion_r1281449756) comment. Then it's hidden away

File: data_model/src/permission.rs
Comment (Arjentix) [reply data_model/src/permission.rs:103]: The main point is `Serialize` and `Deserialize` derives. I'd like it to be consistent with other json in `genesis.json`. It will be weird if it will be string in json format inside another json

Comment (Arjentix) [reply data_model/src/permission.rs:103]: I'll change the name

File: client/tests/integration/smartcontracts/validator_with_custom_token/src/lib.rs
Comment (Arjentix) [reply client/tests/integration/smartcontracts/validator_with_custom_token/src/lib.rs:121]: > it feels like we're tied to serde_json

Yes, as well, as we tied to `parity_scale_codec`. But I don't see much problems here.
The problem of keeping `String` is that it will be Serialized/Deserialized as string, not as JSON. See my response for Roman's comment

File: tools/kagami/src/genesis.rs
Comment (Arjentix) [reply tools/kagami/src/genesis.rs:178]: When how do we say to which transaction we append instruction? Genesis contains transactions, not instructions directly

File: wasm/validator/Cargo.toml
Comment (Arjentix) [reply wasm/validator/Cargo.toml:22]: Indeed, we don't

File: wasm/validator/src/permission.rs
Comment (Arjentix) [reply wasm/validator/src/permission.rs:53]: You're right, we don't

File: data_model/src/permission.rs
Comment (Erigara) [reply data_model/src/permission.rs]: > The main point is `Serialize` and `Deserialize` derives. I'd like it to be consistent with other json in `genesis.json`. It will be weird if it will be string in json format inside another json

We can provide our own `Serialize` and `Deserialize` which will do conversion `Serialized` -> `serde_json::Value` -> `String` (and wise versa). What do you think?

File: tools/kagami/src/genesis.rs
Comment (mversic) [reply tools/kagami/src/genesis.rs]: as far as I am aware it always contains only single transaction

File: client/tests/integration/smartcontracts/validator_with_custom_token/src/lib.rs
Comment (mversic) [reply client/tests/integration/smartcontracts/validator_with_custom_token/src/lib.rs]: let's move discussion to that comment

File: tools/kagami/src/genesis.rs
Comment (mversic) [reply tools/kagami/src/genesis.rs]: and in case there are multiple why exactly would we change specifically first transaction of all transactions?

Comment (Arjentix) [reply tools/kagami/src/genesis.rs]: > as far as I am aware it always contains only single transaction

Yes, we just always use corner case.

> and in case there are multiple why exactly would we change specifically first transaction of all transactions?

I think there is no real reason.

But having `append_instruction()` in `genesis` does not tell me that it will append instruction to the latest transaction.

File: data_model/src/permission.rs
Comment (Arjentix) [reply data_model/src/permission.rs]: And what's the point of doing that if it would be the same, we just move conversion from `Encode/Decode` to `Serialize/Deserialize`?

While having `serde_json::Value` protects us from currupting json with out actions, which `String` cannot provide

File: data_model/src/permission.rs
Comment (0x009922) [inline data_model/src/permission.rs:135]: Encoding JSON twice might be expensive. Although it is not critical now.

**Review by 0x009922 — APPROVED** [review]

File: data_model/src/permission.rs
Comment (Erigara) [reply data_model/src/permission.rs]: > And what's the point of doing that if it would be the same, we just move conversion from Encode/Decode to Serialize/Deserialize?

We can validate that string which we received is valid json without actually parsing it:

```rust
let _: serde::de::IgnoredAny = serde_json::from_str(data)?;
```

> While having serde_json::Value protects us from currupting json with our actions, which String cannot provide

Which actions may corrupt permission token payload?

And we anyway have to deserialize/decode permission token inside validator, so we can verify that we received valid json.

File: tools/kagami/src/genesis.rs
Comment (mversic) [reply tools/kagami/src/genesis.rs]: ok, it's not related to this PR anyhow

File: data_model/src/permission.rs
Comment (Arjentix) [reply data_model/src/permission.rs:135]: Will address this under [this](https://github.com/hyperledger/iroha/pull/3775#discussion_r1287191647) discussion

File: data_model/src/permission.rs
Comment (Arjentix) [reply data_model/src/permission.rs]: Check out new code

**Review by Erigara — APPROVED** [review]

**Review by 0x009922 — DISMISSED** [review]

**Review by 0x009922 — APPROVED** [review]

**Comment by Arjentix** [comment]
@mversic , waiting for your approval

**Review by mversic — APPROVED** [review]

**Pull request merged**
